ruby-test-mode.el

```diff
@@ -69,49 +70,91 @@ Test Driven Development in Ruby."
+    (,(pcre-to-elisp "(.*)/spec/lib/(.*)_spec\\.rb$") "\\1/lib/\\2.rb")
+    (,(pcre-to-elisp "(.*)/test/lib/(.*)_test\\.rb$") "\\1/lib/\\2.rb")
+
+    ;; Project/spec/aaa/bbb_spec.rb => Project/lib/aaa/bbb.rb
+    (,(pcre-to-elisp "(.*)/spec/(.*)_spec\\.rb$") "\\1/lib/\\2.rb")
+
+    (,(pcre-to-elisp "(.*)/test/unit/(.*)_test\\.rb$") "\\1/app/models/\\2.rb" "\\1/lib/\\2.rb")
+    (,(pcre-to-elisp "(.*)/test/functional/(.*)_test\\.rb$") "\\1/app/controllers/\\2.rb")
+
+    ;; Project/test/aaa/bbb_test.rb => Project/lib/aaa/bbb.rb
+    (,(pcre-to-elisp "(.*)/test/(.*)_test\\.rb$") "\\1/lib/\\2.rb")
+
+    ;; make ruby-test-mode support asserts spec.
+    ;; (,(pcre-to-elisp "(.*)/spec/javascripts/(.*)_spec\\.(js|coffee)$") "\\1/app/assets/\\2.\\3")
+
+    ;; in same folder,  gem/aaa_spec.rb => gem/aaa.rb
+    (,(pcre-to-elisp "(.*)_(spec|test)\\.rb$") "\\1.rb")
+    )
   "Regular expressions to map Ruby implementation to unit
 filenames). The first element in each list is the match, the
 second the replace expression."
   :type '(list)
   :group 'ruby-test)
 
 (defcustom ruby-test-specification-filename-mapping
-  '(
-    ("\\(.*\\)\\(app/\\)\\(.*\\)\\([^/]*\\)\\(\\.rb\\)$" "\\1spec/\\3\\4_spec\\5" "\\1test/\\3\\4_test\\5")
-    ("\\(.*\\)\\(app/views\\)\\(.*\\)$" "\\1spec/views\\3_spec.rb")
-    ("\\(.*?\\)\\(lib/\\)\\(.*\\)\\([^/]*\\)\\(\\.rb\\)$" "\\1spec/\\2\\3\\4_spec\\5")
-    ("\\(.*\\)\\(\\.rb\\)$" "\\1_spec\\2"))
+  `(
+    (,(pcre-to-elisp "(.*)/config/routes\\.rb$") "\\1/spec/routing/routes_spec.rb" "\\1/test/routing/routes_test.rb")
+    (,(pcre-to-elisp "(.*)/app/views/(.*)$") "\\1/spec/views/\\2_spec.rb" "\\1/test/views/\\2_test.rb")
+
+    ;; everything in app, should exist same path in spec/test.
+    (,(pcre-to-elisp "(.*)/app/(.*)\\.rb$") "\\1/spec/\\2_spec.rb" "\\1/test/\\2_test.rb")
+
+    (,(pcre-to-elisp "(.*)/lib/(tasks/.*)\\.rake$") "\\1/spec/\\2_tasks_spec.rb" "\\1/test/\\2_tasks_test.rb")
+
+    ;; Project/lib/aaa/bbb.rb, search order:
+    ;; => Project/spec/lib/aaa/bbb_spec.rb, Project/spec/aaa/bbb_spec.rb
+    (,(pcre-to-elisp "(.*)/lib/(.*)\\.rb$") "\\1/spec/\\2_spec.rb" "\\1/spec/lib/\\2_spec.rb" "\\1/test/\\2_test.rb" "\\1/test/lib/\\2_test.rb")
+
+    ;; make ruby-test-mode support asserts spec.
+    ;; (,(pcre-to-elisp "(.*)/app/assets/(.*)\\.(js|coffee)$") "\\1/spec/javascripts/\\2_spec.\\3" "\\1/spec/js/\\2_test.\\3")
+
+    ;; in same folder,  gem/aaa.rb => gem/aaa_spec.rb
+    (,(pcre-to-elisp "(.*)\\.rb$") "\\1_spec.rb" "\\1_test.rb")
+    )
   "Regular expressions to map Ruby specification to
 implementation filenames). The first element in each list is the
 match, the second the replace expression."
   :type '(list)
   :group 'ruby-test)
 
+;; TODO: It seem like we does not need this mapping anymore.
+;; We could add more candicaate to `ruby-test-specification-filename-mapping' to
+;; instead this.
 (defcustom ruby-test-unit-filename-mapping
```

Comment:
> If this is not needed anymore, I think ruby-test-unit-filename-mapping and ruby-test-specification-filename-mapping should be merged into one. Can't remember exactly why we had 2 of those.

I read the code again.
I think the purpose is:  if not exist test file, use those map to decide what file we need to create.
This will give a change to create a `???_test.rb` or `???_spec`.  choice which decide by the 
variable `ruby-test-default-library`.

I think this is not need, we can deicde this according the test directory name.